### ISSUE TITLE
Add i18n to Remote Desktop

### ISSUE THREAD

**Issue by blairdrummond** [StatCan/kubeflow-containers-desktop#46]
See if we can configure bilingualism options in the remote desktop image.

**Comment by Jose-Matsuda** [comment]
Looked into this while waiting for my image to build. 
Using one of my older images for testing Jupyter which had the Lang env variables set to French, as well as the French locale being installed I found that most of it was able to be internationalized with video demonstrating some of the options [here](https://streamable.com/gw97ad).

I've taken a look through [what we base this proj on](https://github.com/ml-tooling/ml-workspace#desktop-gui) as well as the actual [noVNC project](https://github.com/novnc/noVNC) and there's almost nothing on i18n.

Having said that for the noVNC proj there are some [language files](https://github.com/novnc/noVNC/tree/master/po) (no French file, but it is pretty short) and these correspond to a few of the extra messages such as "Connected (encrypted) to" "Failed to connect to server" etc. Users would probably run into around 10 of the translations that are provided by the po file. 

This matches with what Nancy was thinking of doing before she ran into issues with running and testing her changes

IF this approach is wanted, it is very likely a small change that fits with #48 as well as what Wendy has talked to me about PodDefaults and getting a language through there. I still need to check exactly which ENV needs to be set to French or English for the language to be applied but it may be a change in the init.sh to then change `LANG` to `fr_FR.utf8` or `en_US.utf8`  through an env var set through the PodDefault

**Comment by Jose-Matsuda** [comment]
Commenting for future Jose / whoever

Aside from the comment above, to get the vscode instance inside novnc remote desktop to display in French, similar to vscode in browser you need to add an `argv.json` file but the path is now `/home/jovyan/.vscode/argv.json`. If that file is there with say `"locale":"fr"` then upon starting up you will be greeted with vscode, _but in French_

7/12/2020: This file does not matter. It gets covered by the `export LANG..etc` phrases. The user can still change the display language to English if they want. But the default is determined by the `start-vnc-server` file where once a user hits the landing page, the browser language is checked, if it does not match 'en' then vnc needs to be killed, and then restarted to get to French.

**Comment by Jose-Matsuda** [comment]
Similar to other implementations of i18n in remote desktop, to have the noVNC change language (as seen in the video above) while the user is connected is to have some sort of `export LC_ALL=fr_FR` or en_EN in the [start-vnc-server](https://github.com/StatCan/kubeflow-containers-desktop/blob/master/base/resources/scripts/start-vnc-server.sh#L13). This would likely again read from the `argv.json` file used by vscode. 
The service would just need to be killed when the browser language and a refresh on the landing page followed by an auto-restart courtesy of supervisord. The 'kill' of the service is a shell command of `kill $(lsof -t -i:5901)` where5901 is the port that the vnc_server runs on. The default signal sent by `kill` is the `SIGTERM` which is `trap`ped by [line 33](https://github.com/StatCan/kubeflow-containers-desktop/blob/master/base/resources/scripts/start-vnc-server.sh#L33) to have the shutdown signal be sent.

**Comment by Jose-Matsuda** [comment]
**All apps** so Vscode, Firefox, LibreOffice Calc, Netdata should be starting in French. Again leverage Wendy's work on passing an environment variable in. For now, test with manually setting environment variables to French and seeing what else needs to be done

VSCODE: Requires the LANG to be fr_FR
Firefox: Requires an if in `firefox.sh` to get the right installation of FF, if LANG == fr_fr... then French install else English
Netdata: No plans for [multilingual support](https://github.com/netdata/netdata/issues/8933) and as noted by the comments, they have moved away from the [localization project](https://github.com/netdata/localization) which is now a read-only repo
LibreOffice: [pretty simple](https://streamable.com/ae5826). just need to add `apt-get install libreoffice-help-fr libreoffice-l10n-fr` after the `apt-get install libreoffice...` area and the locales will take care of it. I have made the locales "english" while keeping the `apt-get install` line in and it was English (and with locales being French it would start in French)

**Comment by Jose-Matsuda** [comment]
Programs which **have translations** and I know how to access
- VsCode: requires LANG to be fr_FR
- Firefox: requires using the **French** installation of FF, can use the LANG env variable
- LibreOffice: see above

Programs which have translations **but I don't know how to leverage them yet**
These are essentially the programs whose projects have a `.po` file for the French langauge
- thunar file manager+ bulk rename: https://gitlab.xfce.org/xfce/thunar
- xcfe4-terminal 0.8.7.4: https://gitlab.xfce.org/apps/xfce4-terminal
- Baobab 3.28.0: graphical tool to analyze disk usage https://gitlab.gnome.org/GNOME/baobab
- gFTP:  https://github.com/masneyb/gftp
- GDebi 0.9.5.7: (install and view software packages) https://launchpad.net/gdebi leading to https://bazaar.launchpad.net/~gdebi-developers/gdebi/trunk/files
- Task Manager 1.2.0 xfce : https://gitlab.xfce.org/apps/xfce4-taskmanager
- gigolo 0.4.2: (simple frontend to easily connect/mount to local and remote filesystems) https://gitlab.xfce.org/apps/gigolo
- Eye of GNOME image viewer (3.28.1) : https://gitlab.gnome.org/GNOME/eog
- xarchiver 0.5.4.12 (GTK+ only archive manager) : abandoned, but has a PO https://sourceforge.net/p/xarchiver/code/HEAD/tree/trunk/
- Catfish 1.4.4 (versatile file searching tool):  https://gitlab.xfce.org/apps/catfish
- Font Manager 0.7.3 (Simple font management for GTK+ desktop environments): https://github.com/FontManager/font-manager PO HERE


Programs which **do not have translations**
- Netdata: see above comment, no planned support
- glogg: log explorer, has not been updated in 2 years [a PR adding Chinese has 0 activity](https://github.com/nickbnf/glogg/pull/276)

- Some extra messages which currently do not have FR translations [here](https://github.com/StatCan/kubeflow-containers-desktop/blob/master/base/resources/novnc/app/ui.js#L1791). Would need to create and put this json under resources/novnc/app/locale/fr.json in the container

**Comment by Jose-Matsuda** [comment]
Attempting to set the locales of the ENV at a deeper level because when I replaced the en_US with fr_FR some configuration settings stayed "English"

The item that I thought was the big cheese was when you ran `debconf-show locales` and the result would be 
![image](https://user-images.githubusercontent.com/23174198/103558918-5d6c6800-4e83-11eb-9bf5-dd0bbf2dcde5.png)
"None" for default_environment_locale

However, even after getting it to show up as "fr_CA" applications continued to stay in English
![image](https://user-images.githubusercontent.com/23174198/103559109-9e647c80-4e83-11eb-9b77-dd1485c88818.png)

As well as looking at /etc/locale.gen where fr_CA is uncommented out
![image](https://user-images.githubusercontent.com/23174198/103671012-58bdb780-4f48-11eb-84cc-56e780f432fa.png)
and of course in /etc/default/locale
![image](https://user-images.githubusercontent.com/23174198/103671130-8145b180-4f48-11eb-8fa1-63aa8f1667d6.png)


Sources tried were [wiki.debian](https://wiki.debian.org/Locale#Standard) (did not try the optional part). Looked again at the SO post in the [code](https://stackoverflow.com/questions/28405902/how-to-set-the-locale-inside-a-debian-ubuntu-docker-container/38553499#38553499), confirming checking locale in the [env and going through files](https://serverfault.com/a/54599)

**Comment by Jose-Matsuda** [comment]
Set up my own xfce4 in a vm and was able to get it into French by selecting "French" at install and it modified the two files I expected. 
![image](https://user-images.githubusercontent.com/23174198/103673067-22ce0280-4f4b-11eb-92b6-462e81245638.png)
still has the application menu in French
![image](https://user-images.githubusercontent.com/23174198/103673198-501ab080-4f4b-11eb-9369-e8fd5f72d3e3.png)
Locales
![image](https://user-images.githubusercontent.com/23174198/103674755-1b0f5d80-4f4d-11eb-9fbd-d30fae68f1f4.png)


Grasping a bit and looked at [novnc](https://github.com/novnc/noVNC/tree/master/po). I don't think there's anything here unless there's some setting in our setup that has the current novnc installation take precedence over the xfce4 variables in the remote desktop that makes thunar and them not be localized

**Comment by Jose-Matsuda** [comment]
Sprung up a basic image of xcfe based on [x11docker](https://hub.docker.com/r/x11docker/xfce) simply adding the locale. As this [video](https://streamable.com/dmm6d9) shows, at least with no outside interference xcfe.

**Comment by Jose-Matsuda** [comment]
Update, it appears that the localization will work on **Ubuntu:16.04** but will not on **Ubuntu:18.04**. I found this out after checking Blair's image (which uses 16.04) with the current "locale" setting method and then trying it with the exact same image except using ubuntu:18.04.

Upon suggestion by Blair/Brendan will check about gtk2, gtk3, check out ubuntu:20.04 (should be updating anyways), KDE

**Comment by Jose-Matsuda** [comment]
Ok, well gtk2 support is dropped in [xfce 4.16](https://www.xfce.org/download/changelogs/4.16) so for now I'm going to try and update to that version (as well as Ubuntu 20.04) in my own smaller remote desktop project (goodbye insanely long build times) and see if that resolves it first.
Edit: at least on ubuntu packages, for 20.04 its just version [4.14](https://packages.ubuntu.com/focal/xfce4) for now. 4.16 was just released at the end of last year. Ubuntu 18 also only has packages up to 4.12

**Comment by Jose-Matsuda** [comment]
Have a way to get something like _Thunar_ translated for sure. Just to try out I copied over an `mo` file from my `ubuntu:16.04` installation using Blair's image placing it in the `/usr/share/locale-langpack/fr/LC_MESSAGES` folder  
![image](https://user-images.githubusercontent.com/23174198/105043407-7c4a2d00-5a33-11eb-8170-d8949c72020a.png)

The `mo` file can also be placed in the `/usr/share/locale/fr_LC_MESSAGES` folder (in fact the `mo` file for tigervnc is here) and it would have the same effect.

**Comment by Jose-Matsuda** [comment]
Also wild, I can use `msgfmt messages.po -o messages.mo` for some applications that for whatever reason did **not** generate a `mo` file in my `ubuntu:16.04` (in my case it was gdebi).
I took the `po` file from the gdebi project listed above, downloaded the `fr.po`, ran it through doing `msgfmt fr.po -o gdebi.mo` and copied it over and bam I have gdebi in French. 
![image](https://user-images.githubusercontent.com/23174198/105048323-6ccde280-5a39-11eb-8cb5-54e4690938ec.png)

 
The total size of an `mo` folder containing the translations would be around 2 mb.

**Comment by Jose-Matsuda** [comment]
**Regarding the r and geomatics images**
Rstudio here is in a similar spot where the UI is **not** translated, only the console is.
Geomatics with QGIS will work if the environment variables are set to French
![image](https://user-images.githubusercontent.com/23174198/105089930-c18a5100-5a6b-11eb-912e-4d2717134ebc.png)

**Comment by Jose-Matsuda** [comment]
Few outstanding areas of note related to these applications; Baobab (disk usage analyzer), EOG (Eye Of Gnome image viewer), DB Browser for SQLite.

For DB Browser, just the **title** is untranslated, everything else is
![image](https://user-images.githubusercontent.com/23174198/105378014-61192200-5bd9-11eb-8c7a-9f42c245f232.png)
IMO this is fine as it is the _name_ of the application. 

Baobab and EOG have the same issue where again the application title is in English only in the access area. 
![image](https://user-images.githubusercontent.com/23174198/105379303-b1dd4a80-5bda-11eb-9d4d-7ac24f314a84.png)
Note that the descriptions for these two applications in the "Search for applications" area here is in English as well.
I have the same feeling here as from above.  

Things I have tried:
I have tried installing from a ppa repo to find a more "up to date" version but it did not help, I have tried giving the applications the most up to date `mo` file and again no luck. 


These are **very easy to remove** so if its not wanted gone.
